c7n/mu.py

```diff
@@ -1260,6 +1265,129 @@ def remove(self, func, func_deleted=True):
             return True
 
 
+class EventBridgeScheduleSource(AWSEventBase):
+    """
+    Invoke Lambda functions via EventBridge Scheduler.
+    """
+
+    client_service = 'scheduler'
+
+    def get(self, schedule_name, group_name='default'):
+        return resource_exists(self.client.get_schedule,
+                               Name=schedule_name,
+                               GroupName=group_name)
+
+    @staticmethod
+    def delta(src, tgt):
+        """Given two schedules determine if the configuration is the same.
+
+        Name is already implied.
+        """
+        for k in ['State', 'StartDate', 'EndDate', 'ScheduleExpression',
+                  'ScheduleExpressionTimezone', 'Description', 'GroupName']:
+            if src.get(k) != tgt.get(k):
+                return True
+
+        for k in ['Arn', 'RoleArn']:
+            if src.get('Target', {}).get(k) != tgt.get('Target', {}).get(k):
+                return True
+
+        return False  # pragma: no cover
+
+    def __repr__(self):
+        return (f'<CWEvent Type:{self.data.get("type")} Events:'
+                f'{", ".join(map(str, self.data.get("events", [])))}>')
+
+    def add(self, func):
+        params = dict(
+            Name=func.event_name,
+            Description=func.description,
+            State='ENABLED',
+            FlexibleTimeWindow={'Mode': 'OFF'},
+            ScheduleExpression=self.data.get('schedule'),
+            ScheduleExpressionTimezone=self.data.get('timezone', 'Etc/UTC'),
+            GroupName=self.data.get('group-name', 'default'),
+            Target={
+                'Arn': func.arn,
+                'RoleArn': self.data.get('scheduler-role')
+            }
+        )
+
+        schedule = self.get(func.event_name, params['GroupName'])
+
+        if schedule and self.delta(schedule, params):
+            log.debug(f'Updating schedule for {func.event_name} in group {params["GroupName"]}')
+            self.client.update_schedule(**params)
+        elif not schedule:
+            log.debug(f'Creating schedule for {self} in group {params["GroupName"]}')
+            self.client.create_schedule(**params)
+
+        # check for schedules with same func.event_name in any other group in case
+        # group_name has been changed
+        # there should only be one schedule across all groups for each policy lambda
+        self.remove_from_groups(self.client, func, params['GroupName'])
```

Comment:
it feels like the group change should be a delta of the tag in actual function vs the configured function which alleviate having to do api calls to trawl through all schedule groups to check for the function schedule.

Reply:
Thanks for spotting that. That was from before I added the tag with the group name in it (needed for mugc). I'll refactor this to leverage that tag rather than scanning all the groups.

Reply:
haven't had a chance to read fully, traveling today, but I was thinking we could encapsulate via a provision event/facade abstraction with the old lambda state, the current lambda config/object, that we could propagate to event provisioners from lambda manager on publish which would provide the needed context to avoid the api calls which the current interface lacks.

Reply:
That's basically what my long _verbage_ was attempting to describe. :-D

I've been working on an implementation of this. This is how it currently works.

In `LambdaManager`, `publish` calls `_create_or_update` which gets the current config if it exists (not a new function) before updating the function. Return that config to `publish` alongside the existing values going back. Extract the old group name from the tag if it exists with a regex. `publish` now has the previous function config available for any other purpose in the future.

`publish` then calls `func.get_events` where func is a `PolicyLambda` or `LambdaFunction` instance and iterates through the returned list of event objects. The `PolicyLambda` variety is where the event source objects get created. So we pass the old group name (or None) through to `get_events` and if we're creating an `EventBridgeScheduleSource` we pass it on in the __init__ and store as a property. We can now delete an old sched when needed during the call to the `add` method.

I've tried this out and it works without needing to modify any other tests. It results in a decrease of about 15% in the number test json files for the two new tests (mostly in `test_eb_scheduler`).

I'll push a commit shortly with this implementation for you to have a look at when you get time. Thanks for your help refining this feature. Safe travels.